clientconn.go

```diff
@@ -840,6 +840,73 @@ func (ac *addrConn) connect() error {
 	ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
 	ac.mu.Unlock()
 
+	common := func() bool {
+		ac.mu.Lock()
+		if ac.state == connectivity.Shutdown {
+			ac.mu.Unlock()
+			return true
+		}
+
+		ac.state = connectivity.TransientFailure
+		ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
+		ac.cc.resolveNow(resolver.ResolveNowOption{})
+		ac.curAddr = resolver.Address{}
+		ac.mu.Unlock()
+
+		if err := ac.resetTransport(); err != nil {
+			ac.mu.Lock()
+			ac.printf("transport exiting: %v", err)
+			ac.mu.Unlock()
+			grpclog.Warningf("grpc: addrConn.transportMonitor exits due to: %v", err)
+			if err != errConnClosing {
+				// Keep this ac in cc.conns, to get the reason it's torn down.
+				ac.tearDown(err)
+			}
+			return true
+		}
+		return false
+	}
+
+	var timer *time.Timer
+	var cdeadline <-chan time.Time
+
+	onDeadline := func() bool {
+		ac.mu.Lock()
+		if ac.backoffDeadline.IsZero() {
+			ac.mu.Unlock()
+			return true
+		}
+		ac.mu.Unlock()
+		err := ac.transport.Close()
+		if err != nil {
+			grpclog.Error(err)
+		}
+
+		return common()
+	}
+
+	onGoAway := func() bool {
+		if timer != nil {
+			timer.Stop()
+		}
+		ac.adjustParams(ac.transport.GetGoAwayReason())
+		return common()
+	}
+
+	onError := func() bool {
+		if timer != nil {
+			timer.Stop()
+		}
+		// If a GoAway happened, regardless of error, adjust our keepalive
```

Comment:
Could you clarify this? I believe the transport does not call onGoAway; only below in the for loop do we call onGoAway. Since the transport is destroy and re-created during resetTransport (inside ac.reconnect):

- One of the two cases (onGoAway, onError) occurs. Let's assume onError
- onError calls ac.reconnect, who calls ac.resetTransport. Transport is recreated (old `ac.transport.GoAway()` is blown away with it, no matter if something wrote to it)
- For loop on clientconn.go:911 wraps around again
- Brand new transport with no GoAway or errors on it, so goAway logic does not happen to the old transport